nucliadb/src/nucliadb/search/search/results_hydrator/base.py

```diff
@@ -0,0 +1,215 @@
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+import asyncio
+import logging
+from typing import Optional
+
+from pydantic import BaseModel
+
+from nucliadb.common.external_index_providers.base import QueryResults as ExternalIndexQueryResults
+from nucliadb.common.external_index_providers.base import TextBlockMatch
+from nucliadb.common.maindb.driver import Transaction
+from nucliadb.common.maindb.utils import get_driver
+from nucliadb.ingest.serialize import managed_serialize
+from nucliadb.search.search import paragraphs
+from nucliadb.search.search.cache import get_resource_cache
+from nucliadb_models.common import FieldTypeName
+from nucliadb_models.resource import ExtractedDataTypeName
+from nucliadb_models.search import (
+    SCORE_TYPE,
+    FindField,
+    FindParagraph,
+    FindResource,
+    KnowledgeboxFindResults,
+    ResourceProperties,
+    TextPosition,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceHydrationOptions(BaseModel):
+    """
+    Options for hydrating resources.
+    """
+
+    show: list[ResourceProperties] = []
+    extracted: list[ExtractedDataTypeName] = []
+    field_type_filter: list[FieldTypeName] = []
+
+
+class TextBlockHydrationOptions(BaseModel):
+    """
+    Options for hydrating text blocks (aka paragraphs).
+    """
+
+    pass
+
+
+async def hydrate_external(
+    retrieval_results: KnowledgeboxFindResults,
+    query_results: ExternalIndexQueryResults,
+    kbid: str,
+    resource_options: ResourceHydrationOptions = ResourceHydrationOptions(),
+    text_block_options: TextBlockHydrationOptions = TextBlockHydrationOptions(),
+    text_block_min_score: Optional[float] = None,
+    max_parallel_operations: int = 50,
+) -> None:
+    """
+    Hydrates the results of an external index retrieval. This includes fetching the text for the text blocks
+    and the metadata for the resources.
+
+    Parameters:
+    - retrieval_results: The results of the retrieval to be hydrated.
+    - query_results: The results of the query to the external index.
+    - kbid: The knowledge base id.
+    - resource_options: Options for hydrating resources.
+    - text_block_options: Options for hydrating text blocks.
+    - max_parallel_operations: The maximum number of hydration parallel operations to perform.
+    """
+    hydrate_ops = []
+    semaphore = asyncio.Semaphore(max_parallel_operations)
+    extracted_text_cache = paragraphs.ExtractedTextCache()
+    rcache = get_resource_cache(clear=True)
+    try:
+        resource_ids = set()
+        for text_block in query_results.iter_matching_text_blocks():
+            if text_block_min_score is not None and text_block.score < text_block_min_score:
+                # Ignore text blocks with a score lower than the minimum
+                continue
```

Comment:
I guess this is from the original implementation, but isn't this hiding a bug in the index?

Reply:
Not really! PInecone doesn't have a way to filter by min_score, so we need to do it on our end